deps: Bump lief to 0.16.6

davidgumberg opened the pull request:
Partially resolves https://github.com/bitcoin/bitcoin/issues/30520, updating `lief` to 0.16.6.

davidgumberg committed:
deps: bump lief to 0.16.6

Some of the primary changes are:
- lief.EXE_FORMATS became lief.Binary.FORMATS IN 0.14.0
        - https://github.com/lief-project/LIEF/blob/494f116c6b8c9163fc4b5f6497233f18424981ee/doc/sphinx/changelog.rst?plain=1#L702
- lief.ARCHITECTURES became lief.Header.ARCHITECTURES in 0.16.0
        - https://github.com/lief-project/LIEF/blob/494f116c6b8c9163fc4b5f6497233f18424981ee/doc/sphinx/changelog.rst?plain=1#L226C18-L227C18
- lief.ELF.ARCH.x86_64 became lief.ELF.ARCH.X86_64

This commit includes a workaround for the bug fixed in
https://github.com/lief-project/LIEF/pull/1218, but the workaround can
be kept, since it makes `has_nx` checks stricter by enforcing both heap
and stack are non-executable.

This change also requires a patch to partially revert a commit to LIEF
(https://github.com/lief-project/LIEF/commit/f23ced2f4ffc170d0a6f40ff4a1bee575e3447cf)
which broke compatibility with versions of scikit-build-core <= 0.10.x.

This patch can be dropped once the guix time machine advances to or
beyond https://codeberg.org/guix/guix/commit/35c5f07e967155d2276c7ec58e5108e4da02c974,
which bumps the scikit-build-core version in guix from 0.9.3 to 0.10.7.

Co-authored-by: willcl-ark <will8clark@gmail.com>
Co-authored-by: fanquake <fanquake@gmail.com>
Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

laanwj commented:
Concept ACK.

> Fixes an existing bug, where bcrypt.dll was not one of the expected symbols in contrib/devtools/symbol-check.py, even though this symbol gets included by way of secp256k1:

From what i understand, that's only used for the example, not for the library itself. The bcrypt-using code (`fill_random` in  `examples/examples_util.h`) is never included in the library.

It's unlikely that the LIEF version changes anything there.

hebasto reviewed (commented):
~Also the LIEF version must be bumped here~

hebasto commented on contrib/guix/manifest.scm:
```diff
@@ -158,7 +158,7 @@ chain for " target " development."))
   (package
     (name "python-lief")
-    (version "0.13.2")
+    (version "0.16.5")
```
Does Guix build succeed?

davidgumberg replied:
In the process of testing this, I'll mark as draft in the meanwhile.

hebasto commented:
Concept ACK.

davidgumberg commented:
> From what i understand, that's only used for the example, not for the library itself. The bcrypt-using code (`fill_random` in `examples/examples_util.h`) is never included in the library.
> 
> It's unlikely that the LIEF version changes anything there.

Oops, my bad, I must have ran `symbol-check.py` against a stale build of the `bcrypt` branch, rebased to drop adding `bcrypt` to `symbol-check.py`, marking draft until I test this as working in Guix.

fanquake commented:
> Add usage notes for contrib/devtools/security-check.py

I think you can actually remove these docs. The symbol / security scripts are not for general developer usage, and there's no expectation that the scripts should pass, outside of a Guix build. We could move them out of contrib to make this more clear.

> Fixes an existing bug, where bcrypt.dll was not one of the expected symbols in contrib/devtools/symbol-check.py, even though this symbol gets included by way of secp256k1:

This seems to be a symptom of the above. Dependencies in libsecp256k1 example code, are not relevant for our release builds. Looking at a Guix build of master, I cannot see any dependency on `bcrypt.dll` in any binary.

davidgumberg commented:
I definitely opened this PR prematurely, thinking I could get away with just bumping the version and fixing the python scripts 😂.

Working on getting guix builds working right, I referenced prior art by @willcl-ark here (https://github.com/willcl-ark/bitcoin/commit/ede9aa93fe6339fe6285c954dc8fbd9ae8623916).

The python build is convinced that a `build` option is set in `config-settings`, even when I override `configure-flags`, which is where `config-settings` is supposed to get it's value from (https://github.com/fanquake/guix/blob/ac2d792aae241f5233ee3fdfa29cd3dbaeb9338c/guix/build/pyproject-build-system.scm#L107-L111) :

```
starting phase `set-pythonpath'
phase `set-pythonpath' succeeded after 0.0 seconds
starting phase `change-directory'
phase `change-directory' succeeded after 0.0 seconds
starting phase `build'
Using 'setup' to build wheels, auto-detected 'setup', override '#f'.
ERROR: Unrecognized options in config-settings:
  build -> Did you mean: build-dir?
error: in phase 'build': uncaught exception:
%exception #<&invoke-error program: "python" arguments: ("-c" "import sys, importlib, json\nconfig_settings = json.loads (sys.argv[3])\nbuilder = importlib.import_module(sys.argv[1])\nbuilder.build_wheel(sys.argv[2], config_settings=config_settings)" "setup" "dist" "[]") exit-status: 7 term-signal: #f stop-signal: #f>
phase `build' failed after 0.1 seconds
command "python" "-c" "import sys, importlib, json\nconfig_settings = json.loads (sys.argv[3])\nbuilder = importlib.import_module(sys.argv[1])\nbuilder.build_wheel(sys.argv[2], config_settings
```

davidgumberg commented on contrib/devtools/security-check.py:
```diff
@@ -247,28 +252,35 @@ def check_MACHO_BRANCH_PROTECTION(binary) -> bool:
     retval: int = 0
     for filename in sys.argv[1:]:
         binary = lief.parse(filename)
+        if binary is None:
```
Yes, but it was to appease the CI linter, I understand now that this script is not meant for general use, so I changed this to just cast to `lief.Binary` without checking

laanwj commented:
> The symbol / security scripts are not for general developer usage, and there's no expectation that the scripts should pass, outside of a Guix build. We could move them out of contrib to make this more clear.

i think this makes sense, especially for the symbol check. The security check may have some value as a more general "binary security check". But moving both to a guix-specific path would be fine with me.

fanquake commented:
> But moving both to a guix-specific path would be fine with me.

See #32458.

willcl-ark commented:
Hey @davidgumberg just getting around to the ping here (sorry).

I have a local manifest which builds 0.16.3 lief: https://gist.github.com/willcl-ark/3ac07722025b696910adba256d813cb6

As with the changes to Lief themselves I began, they def need a second look over, but perhaps it will be of some help.

fanquake commented:
I think it'd be worth investigating any issues with `0.16.5`, rather than deferring. Seeing some build output with this branch, which is unexpected:
```bash
[100%] Built target bitcoin-qt
Checking binary security...
have_gnu_relro:#True and have_bindnow: True
have_gnu_relro:#True and have_bindnow: True
have_gnu_relro:#True and have_bindnow: True
have_gnu_relro:#True and have_bindnow: True
have_gnu_relro:#True and have_bindnow: True
have_gnu_relro:#True and have_bindnow: True
have_gnu_relro:#True and have_bindnow: True
[100%] Built target check-security
```

davidgumberg commented:
>I think it'd be worth investigating any issues with 0.16.5, rather than deferring.

Sounds good, marking the PR as draft again while I investigate.

I've bisected to this commit in `lief`: https://github.com/lief-project/LIEF/commit/f23ced2f4ffc170d0a6f40ff4a1bee575e3447cf

Seems related to scikit-build 0.10 [changing](https://iscinumpy.dev/post/scikit-build-core-0-10/#other-changes) `cmake.targets` to `build.targets`, and defining `build.targets` explains why `build` is defined in [`config_settings`](https://peps.python.org/pep-0517/#config-settings). 

----

>Seeing some build output with this branch, which is unexpected:
> 
> [...]
>

Left some logging in, fixed.

hebasto commented on ci/lint/04_install.sh:
```diff
@@ -38,7 +38,7 @@ python3 --version
 ${CI_RETRY_EXE} pip3 install \
   codespell==2.2.6 \
-  lief==0.13.2 \
+  lief==0.16.4 \
```
```suggestion
  lief==0.16.5 \
```

davidgumberg replied:
Thanks, fixed.

hebasto commented on contrib/guix/manifest.scm:
```diff
@@ -176,19 +178,25 @@ chain for " target " development."))
+                "1wq57xgnvpqpjnld03a095wb0fjl719sqrg51n80dv0lx0868n2z"))))
+    (build-system pyproject-build-system)
+    (native-inputs (list cmake-minimal
+                         python-tomli
```
Is `python-tomli` really necessary?

hebasto reviewed (commented):
> > I think it'd be worth investigating any issues with 0.16.5, rather than deferring.
> 
> Sounds good, marking the PR as draft again while I investigate.
> 
> I've bisected to this commit in `lief`: [lief-project/LIEF@f23ced2](https://github.com/lief-project/LIEF/commit/f23ced2f4ffc170d0a6f40ff4a1bee575e3447cf)
> 
> Seems related to scikit-build 0.10 [changing](https://iscinumpy.dev/post/scikit-build-core-0-10/#other-changes) `cmake.targets` to `build.targets`, and defining `build.targets` explains why `build` is defined in [`config_settings`](https://peps.python.org/pep-0517/#config-settings).
> 
> > Seeing some build output with this branch, which is unexpected:
> > [...]
> 
> Left some logging in, fixed.

This requires bumping the Guix time machine to commit [a786cd333c738dacdf156752b428f572b31382b0](https://codeberg.org/guix/guix/commit/a786cd333c738dacdf156752b428f572b31382b0) or later to ensure the appropriate version of the `python-scikit-build-core` package.

fanquake commented:
> Thanks for the suggested fix @hebasto! Bumping the guix time machine to 

I don't think we need to bump anything here (putting aside the fact that the suggested change breaks the riscv & windows builds, and will make many other changes to the release env). We can just apply a one line patch to LIEF. i.e:
```diff
Partially revert f23ced2f4ffc170d0a6f40ff4a1bee575e3447cf

Restore compat with python-scikit-build-core 0.10.x
Can be dropped when using python-scikit-build-core 0.11.x

--- a/api/python/backend/setup.py
+++ b/api/python/backend/setup.py
@@ -101,12 +101,12 @@ def _get_hooked_config(is_editable: bool) -> Optional[dict[str, Union[str, List[
     config_settings = {
         "logging.level": "DEBUG",
         "build-dir": config.build_dir,
-        "build.targets": config.build.targets,
         "install.strip": config.strip,
         "backport.find-python": "0",
         "wheel.py-api":  config.build.py_api,
         "cmake.source-dir": SRC_DIR.as_posix(),
         "cmake.build-type": config.build.build_type,
+        "cmake.targets": config.build.targets,
         "cmake.args": [
             *config.cmake_generator,
             *config.get_cmake_args(is_editable),
```

davidgumberg commented:
Thanks for the patch @fanquake, I've rebased to include https://github.com/fanquake/bitcoin/tree/scikit_compat, but squashed to avoid a broken commit.

> It seems it [should](https://iscinumpy.dev/post/scikit-build-core-0-10/) be:
> 
> ```
> Restore compat with python-scikit-build-core 0.9.x
> Can be dropped when using python-scikit-build-core 0.10.x
> ```

I see, in theory, it should be sufficient to fast-forward to [35c5f07e967155d2276c7ec58e5108e4da02c974](https://codeberg.org/guix/guix/commit/35c5f07e967155d2276c7ec58e5108e4da02c974), but some other build issue with python-pydantic prevents building lief at that commit hash.

Unfortunately, I think there might be an regression in big endian PPC64 header parsing in LIEF: https://github.com/lief-project/LIEF/issues/1217, but I've added a work-around to `contrib/guix/security-check.py` to handle this.

Leaving this as draft for now until #32634 is merged.

hebasto reviewed (commented):
Tested 38d7664dc1a00e76b75186f4146affa4dc538bae:

```
Checking binary security...
Traceback (most recent call last):
  File "/distsrc-base/distsrc-38d7664dc1a0-powerpc64-linux-gnu/contrib/guix/security-check.py", line 285, in <module>
    for (name, func) in CHECKS[etype][arch]:
KeyError: ARCHITECTURES.UNKNOWN
```

UPD. Oh, I missed https://github.com/lief-project/LIEF/issues/1217.

fanquake commented:
Actually, we can now use 0.16.6: https://github.com/lief-project/LIEF/releases/tag/0.16.6, which includes that change.

davidgumberg committed:
contrib: Re-enable FORTIFY check for RISCV

Skipping this check is no longer necessary since https://github.com/lief-project/LIEF/commit/ab85865f279cf02648018417ec8afa12bd0bef24
resolved the issue.

davidgumberg committed:
refactor: contrib: Move FORTIFY check to BASE_ELF

davidgumberg commented:
- Rebased against master to include https://github.com/bitcoin/bitcoin/pull/32634/commits/dbb2d4c3d54759f8c346882b6f98d26d5bb8e816
- Bumped LIEF to 0.16.6 which resolves https://github.com/lief-project/LIEF/issues/1217
- Added a workaround for https://github.com/lief-project/LIEF/pull/1218, this is acceptable and won't need to be resolved once the upstream patch is merged, since this change tightens our MachO `has_nx` checks to both the stack and heap.
- Added a commit to enable a previously disabled check which an upstream fix has been merged for a long time.

Marking this as ready for review.

<details>
<summary>

### Guix hashes

</summary>

```bash
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
```
```
329bfb5ac252bd35d6a1fac4b7beb96eed963b9c745508f9e2ec27f84eb4d613  guix-build-b72926849b73/output/aarch64-linux-gnu/SHA256SUMS.part
3e03e68e9ba0ccc90da474fcb7608b1f946d3de9edcea2e9166cc6e6ff2f5ffc  guix-build-b72926849b73/output/aarch64-linux-gnu/bitcoin-b72926849b73-aarch64-linux-gnu-debug.tar.gz
c1b7af10db7333a6b01c8b1f1600f4631172f9b52547d62330e75cf76401387f  guix-build-b72926849b73/output/aarch64-linux-gnu/bitcoin-b72926849b73-aarch64-linux-gnu.tar.gz
1b5fa8c89de424267a7b382b2283d9ec4cd2c3c73bb258049ebe273089e78f02  guix-build-b72926849b73/output/arm-linux-gnueabihf/SHA256SUMS.part
bb9781f7cd00f454a77a3f29d28b892494ca32381b5e22f60139015771a5f5f2  guix-build-b72926849b73/output/arm-linux-gnueabihf/bitcoin-b72926849b73-arm-linux-gnueabihf-debug.tar.gz
8ac4ddcd0dcd46667331a5b924182c49330bb9215f20ebc07ec1c99a8bb3d977  guix-build-b72926849b73/output/arm-linux-gnueabihf/bitcoin-b72926849b73-arm-linux-gnueabihf.tar.gz
7fe75d1d0f60c53a1f07eee253f4cc61de0b515e119ac12e77637b90356c3522  guix-build-b72926849b73/output/arm64-apple-darwin/SHA256SUMS.part
75c6694568d93c62a92e8a7e11efb4cc50ad608ab1ff09d297d7eccee8a49c9a  guix-build-b72926849b73/output/arm64-apple-darwin/bitcoin-b72926849b73-arm64-apple-darwin-codesigning.tar.gz
4aba767d08bce80056fd3f5645a4b120f6208903684ad352c70635f1b0e7c353  guix-build-b72926849b73/output/arm64-apple-darwin/bitcoin-b72926849b73-arm64-apple-darwin-unsigned.tar.gz
539a69f884f288a174a09b167a7b16675aba40b2e42fe14ba3930305b7fd7118  guix-build-b72926849b73/output/arm64-apple-darwin/bitcoin-b72926849b73-arm64-apple-darwin-unsigned.zip
7807302d3f1174a46864a564bc230d0bfe7654b4721a29a3359bcc82d8a476e8  guix-build-b72926849b73/output/dist-archive/bitcoin-b72926849b73.tar.gz
ae312d931c3ff090bb6cdcfaefcc55ae706d205a46bf8c570b76e89d0b0a3f73  guix-build-b72926849b73/output/powerpc64-linux-gnu/SHA256SUMS.part
4f5f81b5cfa848c57bb42da462f409dfb3558a229a45d732ce1ff7c2cc98f236  guix-build-b72926849b73/output/powerpc64-linux-gnu/bitcoin-b72926849b73-powerpc64-linux-gnu-debug.tar.gz
6d16058a39e5f5bd5734bcf947cba090232dc1d75f021c31a216b71245ff5d24  guix-build-b72926849b73/output/powerpc64-linux-gnu/bitcoin-b72926849b73-powerpc64-linux-gnu.tar.gz
971e27d2a637e9f51eb1b664c044dace2bdf2e75ff2609409043ac6fdd96c42a  guix-build-b72926849b73/output/riscv64-linux-gnu/SHA256SUMS.part
b711bf4724d4e4cec5b4c89c4d757bd22882cea9fd8602a86fee5f6685f6e8f0  guix-build-b72926849b73/output/riscv64-linux-gnu/bitcoin-b72926849b73-riscv64-linux-gnu-debug.tar.gz
c3b2926b39d530e8f18d0be1ab036f37ed7f0afda5570994ba2b0f063674f6c3  guix-build-b72926849b73/output/riscv64-linux-gnu/bitcoin-b72926849b73-riscv64-linux-gnu.tar.gz
b41d93d5c1f2d27a295bdffa99bac5e1edc46feb2f7da997c8cf41784f929cb5  guix-build-b72926849b73/output/x86_64-apple-darwin/SHA256SUMS.part
154ddf968565e6400a72e94ea673dac8556a64a0e2ade9dc1fd00a3d037b0bf5  guix-build-b72926849b73/output/x86_64-apple-darwin/bitcoin-b72926849b73-x86_64-apple-darwin-codesigning.tar.gz
77bff1e4ade35f561fd1a42d0be05ceec3e7b6b6d85296f030da4729add4ea58  guix-build-b72926849b73/output/x86_64-apple-darwin/bitcoin-b72926849b73-x86_64-apple-darwin-unsigned.tar.gz
8353c0e4eb112ba3f1dcac34f074421789a1ec546f05b1b46a36a1d7d540b349  guix-build-b72926849b73/output/x86_64-apple-darwin/bitcoin-b72926849b73-x86_64-apple-darwin-unsigned.zip
01d6f21c57f7de5ce1723cd6d3cf51040d85b5462a16537b476c6cd270ffeff7  guix-build-b72926849b73/output/x86_64-linux-gnu/SHA256SUMS.part
2279870ee6a8f72a3db8a1bef777b52c785da3ed0da24a56dc8b3f00fac150fe  guix-build-b72926849b73/output/x86_64-linux-gnu/bitcoin-b72926849b73-x86_64-linux-gnu-debug.tar.gz
099790f9f05959e0cdd375db46bc42451b350a2e11088a6f7a5e83e4b7d143b1  guix-build-b72926849b73/output/x86_64-linux-gnu/bitcoin-b72926849b73-x86_64-linux-gnu.tar.gz
dfce99265caf6229b115728ca009a42e91636abe7f2719aea077e9aca81685ce  guix-build-b72926849b73/output/x86_64-w64-mingw32/SHA256SUMS.part
85bd305df55b8bbd0518a1c9c1f0ffe373f0ce00fab08f0ebf63f17ebcc0cbcd  guix-build-b72926849b73/output/x86_64-w64-mingw32/bitcoin-b72926849b73-win64-codesigning.tar.gz
e390a2876cca056286b14c7942fe154e77f790da52570b32096a2ae130a186e2  guix-build-b72926849b73/output/x86_64-w64-mingw32/bitcoin-b72926849b73-win64-debug.zip
4ab469d1b5a82940085cb02362f8036e99a38fb700f01bce36ba824d0a1b65a9  guix-build-b72926849b73/output/x86_64-w64-mingw32/bitcoin-b72926849b73-win64-setup-unsigned.exe
becbb9f15416397bdc42ceba464b7880ad88579f0af1aa33483f357e1c3bcb12  guix-build-b72926849b73/output/x86_64-w64-mingw32/bitcoin-b72926849b73-win64-unsigned.zip
```

hebasto commented:
My Guix build:
```
aarch64
74539281616c591d2b7157152ad5959cb7a68b9505ea3fcb74004fb19df34ca4  guix-build-240fe8754d21/output/aarch64-linux-gnu/SHA256SUMS.part
08a9f2153670c0a07819b73694857f448baffdea276f94e2be6e24732bfd80ef  guix-build-240fe8754d21/output/aarch64-linux-gnu/bitcoin-240fe8754d21-aarch64-linux-gnu-debug.tar.gz
70026671b2e698c28d11426d35da71422b4163b52524a3c4442f7dd2606f1c84  guix-build-240fe8754d21/output/aarch64-linux-gnu/bitcoin-240fe8754d21-aarch64-linux-gnu.tar.gz
376d53e484e2c5a353fb9022256e2ea64b7bff91c2d4c37ad7c0af93ff874923  guix-build-240fe8754d21/output/arm-linux-gnueabihf/SHA256SUMS.part
f14547ebf708ef2bb1eb95189b66311d49accacbcd4812daba51ff29f97a2758  guix-build-240fe8754d21/output/arm-linux-gnueabihf/bitcoin-240fe8754d21-arm-linux-gnueabihf-debug.tar.gz
c18762bf9b59aaa109b0305d32f79a7d80e7907e2e35abdb2d9d6d9d8d05965b  guix-build-240fe8754d21/output/arm-linux-gnueabihf/bitcoin-240fe8754d21-arm-linux-gnueabihf.tar.gz
67cb395211298f632b5c0e1751a0c78e625ab938eb6070f8cc330c2e237222fe  guix-build-240fe8754d21/output/arm64-apple-darwin/SHA256SUMS.part
84f7dc819af3beb04142047789553f873eac29dde82538e58f32e2b1bca066e0  guix-build-240fe8754d21/output/arm64-apple-darwin/bitcoin-240fe8754d21-arm64-apple-darwin-codesigning.tar.gz
65024f81a5925a4f8110f6b242895295fbb7ff086dca1f99c1a16a8007c91bad  guix-build-240fe8754d21/output/arm64-apple-darwin/bitcoin-240fe8754d21-arm64-apple-darwin-unsigned.tar.gz
e7b20667cc7a1472cc4da020cbf66685e7ec3c0b1f1d63d7d8737f04e0bbee55  guix-build-240fe8754d21/output/arm64-apple-darwin/bitcoin-240fe8754d21-arm64-apple-darwin-unsigned.zip
77de042a9cbb81ccf76d547559f3872d0de45e5288092f7b9e6a76850826dbfd  guix-build-240fe8754d21/output/dist-archive/bitcoin-240fe8754d21.tar.gz
297a295a0c01e8d577a62fd972ec0d968891755d17c5a45d50de614564aeb281  guix-build-240fe8754d21/output/powerpc64-linux-gnu/SHA256SUMS.part
34109f1e6f20cca085bcb8a11d4ca7c34c089f6afbbfbcce3cc644f2701f20e2  guix-build-240fe8754d21/output/powerpc64-linux-gnu/bitcoin-240fe8754d21-powerpc64-linux-gnu-debug.tar.gz
e9975759dc6ed1ea2634ad4076f90290370d5aecd987117718482122e215f638  guix-build-240fe8754d21/output/powerpc64-linux-gnu/bitcoin-240fe8754d21-powerpc64-linux-gnu.tar.gz
6b973e0ca6b29574f0f80c4266505de02d6ed858f5ab9098db184fc769783a64  guix-build-240fe8754d21/output/riscv64-linux-gnu/SHA256SUMS.part
dd045635ae141c14c370669fc17b2b5c15d7a7c22354a8b343054a05ef4c2eda  guix-build-240fe8754d21/output/riscv64-linux-gnu/bitcoin-240fe8754d21-riscv64-linux-gnu-debug.tar.gz
6dd007bea1ef72ee94f3028b1abea0692e86056a4b0544134de647c803bb05a8  guix-build-240fe8754d21/output/riscv64-linux-gnu/bitcoin-240fe8754d21-riscv64-linux-gnu.tar.gz
6c3e51a29b64c763f7a529088b8c41e6e6f463eb4a12ffcf09882fce5160cb09  guix-build-240fe8754d21/output/x86_64-apple-darwin/SHA256SUMS.part
4c57a09b1e9cfff512a527c9b40b4dc076383022d4ed46e85f21bfbf963dffa1  guix-build-240fe8754d21/output/x86_64-apple-darwin/bitcoin-240fe8754d21-x86_64-apple-darwin-codesigning.tar.gz
c4a99aa5a5051ae673baa6ba9512cbd3df6795c4f40ca2187925b9ecfeec0b89  guix-build-240fe8754d21/output/x86_64-apple-darwin/bitcoin-240fe8754d21-x86_64-apple-darwin-unsigned.tar.gz
aa620c8eea8c0876e6b404226e8e71b6fbced3f8300433c03f28b4230f5f8241  guix-build-240fe8754d21/output/x86_64-apple-darwin/bitcoin-240fe8754d21-x86_64-apple-darwin-unsigned.zip
5fb402e9988e784b4587fc67a5c3d07729bf33a3b298b87b6c7f77f957105935  guix-build-240fe8754d21/output/x86_64-linux-gnu/SHA256SUMS.part
47dbc6c6f0db5f65a9981055a6d08b12bab425d8563074405a4b5302d3d0223d  guix-build-240fe8754d21/output/x86_64-linux-gnu/bitcoin-240fe8754d21-x86_64-linux-gnu-debug.tar.gz
c3805bcb2165a36729518dab307ff0be303967c7eba7a932002e96ae5e34bbc7  guix-build-240fe8754d21/output/x86_64-linux-gnu/bitcoin-240fe8754d21-x86_64-linux-gnu.tar.gz
44302cce79525b1ed622eed8b616f9aca9009f0769ebcf3b41267cb8e02d0b42  guix-build-240fe8754d21/output/x86_64-w64-mingw32/SHA256SUMS.part
7bb4f33f8cbfd3b8e6923f6acd78fdfe5b1e01c4ae2cf23b2a8cb648cdd8789a  guix-build-240fe8754d21/output/x86_64-w64-mingw32/bitcoin-240fe8754d21-win64-codesigning.tar.gz
194e63fbb72e884b0039082300f7bbe87848472faebc6561f54e43fa05cfb46e  guix-build-240fe8754d21/output/x86_64-w64-mingw32/bitcoin-240fe8754d21-win64-debug.zip
456a05dba1c09f3c451ed820b34da1195c290369e6e634e0810bb87208951dc6  guix-build-240fe8754d21/output/x86_64-w64-mingw32/bitcoin-240fe8754d21-win64-setup-unsigned.exe
be932477dc006a3accaa71d9c7fc02fb5a72d809892c0cb5cffb1ca1792c08f3  guix-build-240fe8754d21/output/x86_64-w64-mingw32/bitcoin-240fe8754d21-win64-unsigned.zip
```

hebasto reviewed (commented):
`ninja` is already [listed](https://github.com/bitcoin/bitcoin/blob/4b1d48a6866b24f0ed027334c6de642fc848d083/contrib/guix/manifest.scm#L15) as a dependency in `manifest.scm`, so patching it out doesn't seem necessary:
```diff
--- a/contrib/guix/manifest.scm
+++ b/contrib/guix/manifest.scm
@@ -167,21 +167,13 @@ chain for " target " development."))
                     (url "https://github.com/lief-project/LIEF")
                     (commit version)))
               (file-name (git-file-name name version))
-              (modules '((guix build utils)))
-              (snippet
-               '(begin
-                  ;; Configure build for Python bindings.
-                  (substitute* "api/python/config-default.toml"
-                    (("(ninja         = )true" all m)
-                     (string-append m "false"))
-                    (("(parallel-jobs = )0" all m)
-                     (string-append m (number->string (parallel-job-count)))))))
               (sha256
                (base32
                 "1pq9nagrnkl1x943bqnpiyxmkd9vk99znfxiwqp6vf012b50bz2a"))
               (patches (search-our-patches "lief-scikit-0-9.patch"))))
     (build-system pyproject-build-system)
     (native-inputs (list cmake-minimal
+                         ninja
                          python-scikit-build-core
                          python-pydantic-core
                          python-pydantic-2))
```

fanquake commented on contrib/guix/security-check.py:
```diff
@@ -206,7 +207,10 @@ def check_NX(binary) -> bool:
     Check for no stack execution
     '''
-    return binary.has_nx
+    if binary.format == lief.Binary.FORMATS.MACHO:
```
Actually, looking at this again, with 0.16.6 we can just use `has_nx`, as that does include both checks?

davidgumberg replied:
Unfortunately `has_nx` only includes stack nx checks in 0.16.6 (and main):

https://github.com/lief-project/LIEF/tree/0.16.6/include/LIEF/MachO/Binary.hpp#L579-L593

```cpp

  /// Check if the binary uses `NX` protection
  bool has_nx() const override {
    return !has_nx_stack();
  }

  /// Return True if the **heap** is flagged as non-executable. False otherwise
  bool has_nx_stack() const {
    return !header().has(Header::FLAGS::ALLOW_STACK_EXECUTION);
  }

  /// Return True if the **stack** is flagged as non-executable. False otherwise
  bool has_nx_heap() const {
    return !header().has(Header::FLAGS::NO_HEAP_EXECUTION);
  }
  ```

It seems like the ELF checks are also stack only (https://github.com/lief-project/LIEF/blob/c678f6bdfb3f3199b36e1f8df25f7432a2725f53/src/ELF/Binary.cpp#L1056-L1068)

```cpp
bool Binary::has_nx() const {
  if (const Segment* gnu_stack = get(Segment::TYPE::GNU_STACK)) {
    return !gnu_stack->has(Segment::FLAGS::X);
  }
  
  return false;
}
```

It seems that on most ELF systems the stack is implicitly not executable, I do find some mentions of a header value `PT_GNU_HEAP`, but I'm a bit confused as to whether or not this ELF header bit is actually [respected](https://gcc.gnu.org/legacy-ml/gcc-help/2012-02/msg00122.html), or implemented anywhere, or if heap executability is determined by the kernel, grepping for it in the kernel yields no result.

hebasto commented:
FWIW, https://codeberg.org/guix/guix/pulls/353 has been landed recently.

davidgumberg commented:
Rebased to address @hebasto and @fanquake feedback.

fanquake commented:
Guix build is broken:
```bash
INFO: Building 0dc123319e3b for platform triple x86_64-linux-gnu:
      ...using reference timestamp: 1748567495
      ...running at most 8 jobs
      ...from worktree directory: '/bitcoin'
          ...bind-mounted in container to: '/bitcoin'
      ...in build directory: '/bitcoin/guix-build-0dc123319e3b/distsrc-0dc123319e3b-x86_64-linux-gnu'
          ...bind-mounted in container to: '/distsrc-base/distsrc-0dc123319e3b-x86_64-linux-gnu'
      ...outputting in: '/bitcoin/guix-build-0dc123319e3b/output/x86_64-linux-gnu'
          ...bind-mounted in container to: '/outdir-base/x86_64-linux-gnu'
      ADDITIONAL FLAGS (if set)
          ADDITIONAL_GUIX_COMMON_FLAGS: 
          ADDITIONAL_GUIX_ENVIRONMENT_FLAGS: 
          ADDITIONAL_GUIX_TIMEMACHINE_FLAGS: 
The following derivation will be built:
  /gnu/store/n89gm7ajxy119f9jwhn143awv34r5g9f-python-lief-0.16.6.drv

building /gnu/store/n89gm7ajxy119f9jwhn143awv34r5g9f-python-lief-0.16.6.drv...
\ 'build' phasenote: keeping build directory `/tmp/guix-build-python-lief-0.16.6.drv-1'
builder for `/gnu/store/n89gm7ajxy119f9jwhn143awv34r5g9f-python-lief-0.16.6.drv' failed with exit code 1
build of /gnu/store/n89gm7ajxy119f9jwhn143awv34r5g9f-python-lief-0.16.6.drv failed
View build log at '/var/log/guix/drvs/n8/9gm7ajxy119f9jwhn143awv34r5g9f-python-lief-0.16.6.drv.gz'.
guix shell: error: build of `/gnu/store/n89gm7ajxy119f9jwhn143awv34r5g9f-python-lief-0.16.6.drv' failed
```
```bash
Using 'setup' to build wheels, auto-detected 'setup', override '#f'.
LIEF Version: 0.16.6
2025-06-03 07:34:40,554 - scikit_build_core - INFO - RUN: /gnu/store/wgafw3jn8c9x99wwvd2jczm012ikhmy9-cmake-minimal-3.24.2/bin/cmake -E capabilities
2025-06-03 07:34:40,593 - scikit_build_core - INFO - CMake version: 3.24.2
*** scikit-build-core 0.9.3 using CMake 3.24.2 (wheel)
2025-06-03 07:34:40,598 - scikit_build_core - INFO - Build directory: /tmp/guix-build-python-lief-0.16.6.drv-0/tmpowk_ezed/build
2025-06-03 07:34:40,599 - scikit_build_core - WARNING - No license files found, set wheel.license-files to [] to suppress this warning
*** Configuring CMake...
2025-06-03 07:34:40,601 - scikit_build_core - DEBUG - SITE_PACKAGES: /gnu/store/vc4llpchzzmfdfgvdz54lh9h8wlyzxp7-python-3.10.7/lib/python3.10/site-packages
2025-06-03 07:34:40,601 - scikit_build_core - DEBUG - Extra SITE_PACKAGES: /gnu/store/vc4llpchzzmfdfgvdz54lh9h8wlyzxp7-python-3.10.7/lib/python3.10/site-packages
2025-06-03 07:34:40,601 - scikit_build_core - DEBUG - Set generator: Ninja
Traceback (most recent call last):
  File "<string>", line 4, in <module>
  File "/tmp/guix-build-python-lief-0.16.6.drv-0/source/api/python/backend/setup.py", line 141, in build_wheel
    return _impl(wheel_directory, config_settings, metadata_directory)
  File "/gnu/store/ys4c56b1f2qd583fyz5xxfn87032i3jd-python-scikit-build-core-0.9.3/lib/python3.10/site-packages/scikit_build_core/build/__init__.py", line 31, in build_wheel
    return _build_wheel_impl(
  File "/gnu/store/ys4c56b1f2qd583fyz5xxfn87032i3jd-python-scikit-build-core-0.9.3/lib/python3.10/site-packages/scikit_build_core/build/wheel.py", line 309, in _build_wheel_impl
    builder.configure(
  File "/gnu/store/ys4c56b1f2qd583fyz5xxfn87032i3jd-python-scikit-build-core-0.9.3/lib/python3.10/site-packages/scikit_build_core/builder/builder.py", line 148, in configure
    local_def = set_environment_for_gen(
  File "/gnu/store/ys4c56b1f2qd583fyz5xxfn87032i3jd-python-scikit-build-core-0.9.3/lib/python3.10/site-packages/scikit_build_core/builder/generator.py", line 119, in set_environment_for_gen
    raise NinjaNotFoundError(msg)
scikit_build_core.errors.NinjaNotFoundError: Ninja is required to build
error: in phase 'build': uncaught exception:
%exception #<&invoke-error program: "python" arguments: ("-c" "import sys, importlib, json\nconfig_settings = json.loads (sys.argv[3])\nbuilder = importlib.import_module(sys.argv[1])\nbuilder.build_wheel(sys.argv[2], config_settings=config_settings)" "setup" "dist" "[]") exit-status: 1 term-signal: #f stop-signal: #f> 
phase `build' failed after 0.3 seconds
command "python" "-c" "import sys, importlib, json\nconfig_settings = json.loads (sys.argv[3])\nbuilder = importlib.import_module(sys.argv[1])\nbuilder.build_wheel(sys.argv[2], config_settings=config_settings)" "setup" "dist" "[]" failed with status 1
```

fanquake commented:
Guix Build:
```bash
9607cec72e511df72d1d07ef850015172d6c2707604f46694c019cd37903c65b  guix-build-74dfa4155037/output/aarch64-linux-gnu/SHA256SUMS.part
30916e57efaa2bfd4265d840cde7049c863c4d35cec9d5e70681e7bc428e2d14  guix-build-74dfa4155037/output/aarch64-linux-gnu/bitcoin-74dfa4155037-aarch64-linux-gnu-debug.tar.gz
c462783e946c5e3e61a6a4ab4be1b8ddb6db9be07194d3241b43a8cc4cbe81f7  guix-build-74dfa4155037/output/aarch64-linux-gnu/bitcoin-74dfa4155037-aarch64-linux-gnu.tar.gz
23b60343ce14152fca5f934a3c57d130d665bd5298028f18c2118f646d53841d  guix-build-74dfa4155037/output/arm-linux-gnueabihf/SHA256SUMS.part
81289e78cd5930543c7dfb529ff6a439606d9d2836b0c5664f6bb5b22acf8bb9  guix-build-74dfa4155037/output/arm-linux-gnueabihf/bitcoin-74dfa4155037-arm-linux-gnueabihf-debug.tar.gz
c7bebaa586850cd715fdf990dae5bdc60272f4aaa8b528254d2c4051710b0082  guix-build-74dfa4155037/output/arm-linux-gnueabihf/bitcoin-74dfa4155037-arm-linux-gnueabihf.tar.gz
df16bd44d340f4a3571045d5983fb80cc278f671cea7cf5a03ddf6fd7316ff3a  guix-build-74dfa4155037/output/arm64-apple-darwin/SHA256SUMS.part
e283c1bb16050f583359172bf239965c0a584d804ac3fac7bf847ab8d604261d  guix-build-74dfa4155037/output/arm64-apple-darwin/bitcoin-74dfa4155037-arm64-apple-darwin-codesigning.tar.gz
bfba93334e7ee53b7cc02f589f24d1b2165494c5a90eeba39b00d8ea70b6fb05  guix-build-74dfa4155037/output/arm64-apple-darwin/bitcoin-74dfa4155037-arm64-apple-darwin-unsigned.tar.gz
da1e5edd2cb9ae303b834990ac1fa2ebacb76baf8f0dac514ba1b1fc90711160  guix-build-74dfa4155037/output/arm64-apple-darwin/bitcoin-74dfa4155037-arm64-apple-darwin-unsigned.zip
6ec26a5e72061b1255f50fbe65d84f1a5c21fe8db7e754c62a279d5fa17961f6  guix-build-74dfa4155037/output/dist-archive/bitcoin-74dfa4155037.tar.gz
6c31f2e067a67646ade357afa58baa2ec213064bd4a9e6735c9f77f012a7e916  guix-build-74dfa4155037/output/powerpc64-linux-gnu/SHA256SUMS.part
9e0b5d5784dc1a5a8b0226d4140edd1259e267e31293e7316ba13aa2f754e142  guix-build-74dfa4155037/output/powerpc64-linux-gnu/bitcoin-74dfa4155037-powerpc64-linux-gnu-debug.tar.gz
51761c0923ca13bee3710ed612e6fb5fedb5b000c183c618c3580085746e8501  guix-build-74dfa4155037/output/powerpc64-linux-gnu/bitcoin-74dfa4155037-powerpc64-linux-gnu.tar.gz
4e827c02af15d7fd152440027f854e324e2e1665279700e34beebef70dca773b  guix-build-74dfa4155037/output/riscv64-linux-gnu/SHA256SUMS.part
aeecf0b201d17a9109f3978be2ba375f505b4057c8c442b0301dba6ddb661c42  guix-build-74dfa4155037/output/riscv64-linux-gnu/bitcoin-74dfa4155037-riscv64-linux-gnu-debug.tar.gz
8e59a5f97edbb701dc3dd9ab89bfe49ae0fd658b22f44ee81dacab5506fdb763  guix-build-74dfa4155037/output/riscv64-linux-gnu/bitcoin-74dfa4155037-riscv64-linux-gnu.tar.gz
fe4d6701e2ace1c4491b216ed37b97f6e89583ed70cad4ad0f2d389a05ecadcf  guix-build-74dfa4155037/output/x86_64-apple-darwin/SHA256SUMS.part
218d28c6e890893edb14474f807278050aef4996aa6b64cd07b8d9021ab64244  guix-build-74dfa4155037/output/x86_64-apple-darwin/bitcoin-74dfa4155037-x86_64-apple-darwin-codesigning.tar.gz
e54ba350d8adf218f99345142d25173e5fbe99783335377d209333772013575b  guix-build-74dfa4155037/output/x86_64-apple-darwin/bitcoin-74dfa4155037-x86_64-apple-darwin-unsigned.tar.gz
e9b4aff1fbbe682a9a1e5cd088cbf741b45b04594a475834455240d1a7b24277  guix-build-74dfa4155037/output/x86_64-apple-darwin/bitcoin-74dfa4155037-x86_64-apple-darwin-unsigned.zip
aed2425d7c72902793e000cd5ef8dd37ff9b026981fd7c60308cd9c04c5ccaed  guix-build-74dfa4155037/output/x86_64-linux-gnu/SHA256SUMS.part
91de1bcc9afa636965676116fa12f1f188a58357aa629656cc6a779a618f8b51  guix-build-74dfa4155037/output/x86_64-linux-gnu/bitcoin-74dfa4155037-x86_64-linux-gnu-debug.tar.gz
7151a6e3d9729a8396b0598fbf0d6f3247ed2687bdb65c1ad0859ed4b7a2f50d  guix-build-74dfa4155037/output/x86_64-linux-gnu/bitcoin-74dfa4155037-x86_64-linux-gnu.tar.gz
2213a8e321ae4777ac5ed0ff1481e6dfc0267939ebdb93b3f5bd522a18888076  guix-build-74dfa4155037/output/x86_64-w64-mingw32/SHA256SUMS.part
6089c80b0c312d10cd27397ffa6dd13ae75f3aa54edbabc7bceacf67c68f640a  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-codesigning.tar.gz
08de160ec02609f934282436790b8642b9765e897d5bb475022f43f57612fd60  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-debug.zip
916508cebfa6cc5007c28afb02ab7dedd730b0caa4f5de913ea5f0c47089deaa  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-setup-unsigned.exe
343e0b9475ca9a836fce21f64b95355148819a767c6bc7c181bb4aa0ab082139  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-unsigned.zip
```

hebasto reviewed (commented):
My Guix build:
```
aarch64
9607cec72e511df72d1d07ef850015172d6c2707604f46694c019cd37903c65b  guix-build-74dfa4155037/output/aarch64-linux-gnu/SHA256SUMS.part
30916e57efaa2bfd4265d840cde7049c863c4d35cec9d5e70681e7bc428e2d14  guix-build-74dfa4155037/output/aarch64-linux-gnu/bitcoin-74dfa4155037-aarch64-linux-gnu-debug.tar.gz
c462783e946c5e3e61a6a4ab4be1b8ddb6db9be07194d3241b43a8cc4cbe81f7  guix-build-74dfa4155037/output/aarch64-linux-gnu/bitcoin-74dfa4155037-aarch64-linux-gnu.tar.gz
23b60343ce14152fca5f934a3c57d130d665bd5298028f18c2118f646d53841d  guix-build-74dfa4155037/output/arm-linux-gnueabihf/SHA256SUMS.part
81289e78cd5930543c7dfb529ff6a439606d9d2836b0c5664f6bb5b22acf8bb9  guix-build-74dfa4155037/output/arm-linux-gnueabihf/bitcoin-74dfa4155037-arm-linux-gnueabihf-debug.tar.gz
c7bebaa586850cd715fdf990dae5bdc60272f4aaa8b528254d2c4051710b0082  guix-build-74dfa4155037/output/arm-linux-gnueabihf/bitcoin-74dfa4155037-arm-linux-gnueabihf.tar.gz
df16bd44d340f4a3571045d5983fb80cc278f671cea7cf5a03ddf6fd7316ff3a  guix-build-74dfa4155037/output/arm64-apple-darwin/SHA256SUMS.part
e283c1bb16050f583359172bf239965c0a584d804ac3fac7bf847ab8d604261d  guix-build-74dfa4155037/output/arm64-apple-darwin/bitcoin-74dfa4155037-arm64-apple-darwin-codesigning.tar.gz
bfba93334e7ee53b7cc02f589f24d1b2165494c5a90eeba39b00d8ea70b6fb05  guix-build-74dfa4155037/output/arm64-apple-darwin/bitcoin-74dfa4155037-arm64-apple-darwin-unsigned.tar.gz
da1e5edd2cb9ae303b834990ac1fa2ebacb76baf8f0dac514ba1b1fc90711160  guix-build-74dfa4155037/output/arm64-apple-darwin/bitcoin-74dfa4155037-arm64-apple-darwin-unsigned.zip
6ec26a5e72061b1255f50fbe65d84f1a5c21fe8db7e754c62a279d5fa17961f6  guix-build-74dfa4155037/output/dist-archive/bitcoin-74dfa4155037.tar.gz
6c31f2e067a67646ade357afa58baa2ec213064bd4a9e6735c9f77f012a7e916  guix-build-74dfa4155037/output/powerpc64-linux-gnu/SHA256SUMS.part
9e0b5d5784dc1a5a8b0226d4140edd1259e267e31293e7316ba13aa2f754e142  guix-build-74dfa4155037/output/powerpc64-linux-gnu/bitcoin-74dfa4155037-powerpc64-linux-gnu-debug.tar.gz
51761c0923ca13bee3710ed612e6fb5fedb5b000c183c618c3580085746e8501  guix-build-74dfa4155037/output/powerpc64-linux-gnu/bitcoin-74dfa4155037-powerpc64-linux-gnu.tar.gz
4e827c02af15d7fd152440027f854e324e2e1665279700e34beebef70dca773b  guix-build-74dfa4155037/output/riscv64-linux-gnu/SHA256SUMS.part
aeecf0b201d17a9109f3978be2ba375f505b4057c8c442b0301dba6ddb661c42  guix-build-74dfa4155037/output/riscv64-linux-gnu/bitcoin-74dfa4155037-riscv64-linux-gnu-debug.tar.gz
8e59a5f97edbb701dc3dd9ab89bfe49ae0fd658b22f44ee81dacab5506fdb763  guix-build-74dfa4155037/output/riscv64-linux-gnu/bitcoin-74dfa4155037-riscv64-linux-gnu.tar.gz
fe4d6701e2ace1c4491b216ed37b97f6e89583ed70cad4ad0f2d389a05ecadcf  guix-build-74dfa4155037/output/x86_64-apple-darwin/SHA256SUMS.part
218d28c6e890893edb14474f807278050aef4996aa6b64cd07b8d9021ab64244  guix-build-74dfa4155037/output/x86_64-apple-darwin/bitcoin-74dfa4155037-x86_64-apple-darwin-codesigning.tar.gz
e54ba350d8adf218f99345142d25173e5fbe99783335377d209333772013575b  guix-build-74dfa4155037/output/x86_64-apple-darwin/bitcoin-74dfa4155037-x86_64-apple-darwin-unsigned.tar.gz
e9b4aff1fbbe682a9a1e5cd088cbf741b45b04594a475834455240d1a7b24277  guix-build-74dfa4155037/output/x86_64-apple-darwin/bitcoin-74dfa4155037-x86_64-apple-darwin-unsigned.zip
aed2425d7c72902793e000cd5ef8dd37ff9b026981fd7c60308cd9c04c5ccaed  guix-build-74dfa4155037/output/x86_64-linux-gnu/SHA256SUMS.part
91de1bcc9afa636965676116fa12f1f188a58357aa629656cc6a779a618f8b51  guix-build-74dfa4155037/output/x86_64-linux-gnu/bitcoin-74dfa4155037-x86_64-linux-gnu-debug.tar.gz
7151a6e3d9729a8396b0598fbf0d6f3247ed2687bdb65c1ad0859ed4b7a2f50d  guix-build-74dfa4155037/output/x86_64-linux-gnu/bitcoin-74dfa4155037-x86_64-linux-gnu.tar.gz
2213a8e321ae4777ac5ed0ff1481e6dfc0267939ebdb93b3f5bd522a18888076  guix-build-74dfa4155037/output/x86_64-w64-mingw32/SHA256SUMS.part
6089c80b0c312d10cd27397ffa6dd13ae75f3aa54edbabc7bceacf67c68f640a  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-codesigning.tar.gz
08de160ec02609f934282436790b8642b9765e897d5bb475022f43f57612fd60  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-debug.zip
916508cebfa6cc5007c28afb02ab7dedd730b0caa4f5de913ea5f0c47089deaa  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-setup-unsigned.exe
343e0b9475ca9a836fce21f64b95355148819a767c6bc7c181bb4aa0ab082139  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-unsigned.zip
```

davidgumberg commented:
```console
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
```

```
9607cec72e511df72d1d07ef850015172d6c2707604f46694c019cd37903c65b  guix-build-74dfa4155037/output/aarch64-linux-gnu/SHA256SUMS.part
30916e57efaa2bfd4265d840cde7049c863c4d35cec9d5e70681e7bc428e2d14  guix-build-74dfa4155037/output/aarch64-linux-gnu/bitcoin-74dfa4155037-aarch64-linux-gnu-debug.tar.gz
c462783e946c5e3e61a6a4ab4be1b8ddb6db9be07194d3241b43a8cc4cbe81f7  guix-build-74dfa4155037/output/aarch64-linux-gnu/bitcoin-74dfa4155037-aarch64-linux-gnu.tar.gz
23b60343ce14152fca5f934a3c57d130d665bd5298028f18c2118f646d53841d  guix-build-74dfa4155037/output/arm-linux-gnueabihf/SHA256SUMS.part
81289e78cd5930543c7dfb529ff6a439606d9d2836b0c5664f6bb5b22acf8bb9  guix-build-74dfa4155037/output/arm-linux-gnueabihf/bitcoin-74dfa4155037-arm-linux-gnueabihf-debug.tar.gz
c7bebaa586850cd715fdf990dae5bdc60272f4aaa8b528254d2c4051710b0082  guix-build-74dfa4155037/output/arm-linux-gnueabihf/bitcoin-74dfa4155037-arm-linux-gnueabihf.tar.gz
df16bd44d340f4a3571045d5983fb80cc278f671cea7cf5a03ddf6fd7316ff3a  guix-build-74dfa4155037/output/arm64-apple-darwin/SHA256SUMS.part
e283c1bb16050f583359172bf239965c0a584d804ac3fac7bf847ab8d604261d  guix-build-74dfa4155037/output/arm64-apple-darwin/bitcoin-74dfa4155037-arm64-apple-darwin-codesigning.tar.gz
bfba93334e7ee53b7cc02f589f24d1b2165494c5a90eeba39b00d8ea70b6fb05  guix-build-74dfa4155037/output/arm64-apple-darwin/bitcoin-74dfa4155037-arm64-apple-darwin-unsigned.tar.gz
da1e5edd2cb9ae303b834990ac1fa2ebacb76baf8f0dac514ba1b1fc90711160  guix-build-74dfa4155037/output/arm64-apple-darwin/bitcoin-74dfa4155037-arm64-apple-darwin-unsigned.zip
6ec26a5e72061b1255f50fbe65d84f1a5c21fe8db7e754c62a279d5fa17961f6  guix-build-74dfa4155037/output/dist-archive/bitcoin-74dfa4155037.tar.gz
6c31f2e067a67646ade357afa58baa2ec213064bd4a9e6735c9f77f012a7e916  guix-build-74dfa4155037/output/powerpc64-linux-gnu/SHA256SUMS.part
9e0b5d5784dc1a5a8b0226d4140edd1259e267e31293e7316ba13aa2f754e142  guix-build-74dfa4155037/output/powerpc64-linux-gnu/bitcoin-74dfa4155037-powerpc64-linux-gnu-debug.tar.gz
51761c0923ca13bee3710ed612e6fb5fedb5b000c183c618c3580085746e8501  guix-build-74dfa4155037/output/powerpc64-linux-gnu/bitcoin-74dfa4155037-powerpc64-linux-gnu.tar.gz
4e827c02af15d7fd152440027f854e324e2e1665279700e34beebef70dca773b  guix-build-74dfa4155037/output/riscv64-linux-gnu/SHA256SUMS.part
aeecf0b201d17a9109f3978be2ba375f505b4057c8c442b0301dba6ddb661c42  guix-build-74dfa4155037/output/riscv64-linux-gnu/bitcoin-74dfa4155037-riscv64-linux-gnu-debug.tar.gz
8e59a5f97edbb701dc3dd9ab89bfe49ae0fd658b22f44ee81dacab5506fdb763  guix-build-74dfa4155037/output/riscv64-linux-gnu/bitcoin-74dfa4155037-riscv64-linux-gnu.tar.gz
fe4d6701e2ace1c4491b216ed37b97f6e89583ed70cad4ad0f2d389a05ecadcf  guix-build-74dfa4155037/output/x86_64-apple-darwin/SHA256SUMS.part
218d28c6e890893edb14474f807278050aef4996aa6b64cd07b8d9021ab64244  guix-build-74dfa4155037/output/x86_64-apple-darwin/bitcoin-74dfa4155037-x86_64-apple-darwin-codesigning.tar.gz
e54ba350d8adf218f99345142d25173e5fbe99783335377d209333772013575b  guix-build-74dfa4155037/output/x86_64-apple-darwin/bitcoin-74dfa4155037-x86_64-apple-darwin-unsigned.tar.gz
e9b4aff1fbbe682a9a1e5cd088cbf741b45b04594a475834455240d1a7b24277  guix-build-74dfa4155037/output/x86_64-apple-darwin/bitcoin-74dfa4155037-x86_64-apple-darwin-unsigned.zip
aed2425d7c72902793e000cd5ef8dd37ff9b026981fd7c60308cd9c04c5ccaed  guix-build-74dfa4155037/output/x86_64-linux-gnu/SHA256SUMS.part
91de1bcc9afa636965676116fa12f1f188a58357aa629656cc6a779a618f8b51  guix-build-74dfa4155037/output/x86_64-linux-gnu/bitcoin-74dfa4155037-x86_64-linux-gnu-debug.tar.gz
7151a6e3d9729a8396b0598fbf0d6f3247ed2687bdb65c1ad0859ed4b7a2f50d  guix-build-74dfa4155037/output/x86_64-linux-gnu/bitcoin-74dfa4155037-x86_64-linux-gnu.tar.gz
2213a8e321ae4777ac5ed0ff1481e6dfc0267939ebdb93b3f5bd522a18888076  guix-build-74dfa4155037/output/x86_64-w64-mingw32/SHA256SUMS.part
6089c80b0c312d10cd27397ffa6dd13ae75f3aa54edbabc7bceacf67c68f640a  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-codesigning.tar.gz
08de160ec02609f934282436790b8642b9765e897d5bb475022f43f57612fd60  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-debug.zip
916508cebfa6cc5007c28afb02ab7dedd730b0caa4f5de913ea5f0c47089deaa  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-setup-unsigned.exe
343e0b9475ca9a836fce21f64b95355148819a767c6bc7c181bb4aa0ab082139  guix-build-74dfa4155037/output/x86_64-w64-mingw32/bitcoin-74dfa4155037-win64-unsigned.zip
```

hebasto reviewed (approved):
ACK 74dfa4155037306d79596521a5c38a9e90487331.

fanquake commented on contrib/guix/security-check.py:
```diff
@@ -253,31 +261,31 @@ def check_MACHO_BRANCH_PROTECTION(binary) -> bool:
     retval: int = 0
     for filename in sys.argv[1:]:
-        binary = lief.parse(filename)
+        binary = cast(lief.Binary, lief.parse(filename))
```
Can you clarify why the casting was added; with it removed, I don't see any issues in our CI.

davidgumberg replied:
Fixed, thanks for being persistent on this, not sure why or what I was doing wrong [earlier](https://github.com/bitcoin/bitcoin/pull/32431#discussion_r2080687125) to see this error.

fanquake commented:
Guix Build:
```bash
8ee8f51264bd5f28d838207486f83db69a1dd4d00eeda009f35ef9ebff794d0f  guix-build-4f56c9145a60/output/aarch64-linux-gnu/SHA256SUMS.part
aaf5a45736f062df340bbd434295a087f90897401a8d9e1489f5ad56c4a82fab  guix-build-4f56c9145a60/output/aarch64-linux-gnu/bitcoin-4f56c9145a60-aarch64-linux-gnu-debug.tar.gz
d6b4a08a7e9dad84dcfdc504896cf0b13a4c2c416ef8608903cd542297936f74  guix-build-4f56c9145a60/output/aarch64-linux-gnu/bitcoin-4f56c9145a60-aarch64-linux-gnu.tar.gz
c236ea00cde9621315a84153c0e672c8f9299d93efe379ff82188e2e018167e3  guix-build-4f56c9145a60/output/arm-linux-gnueabihf/SHA256SUMS.part
719c221a599298d5795ce82c0dc051ea512e2f17d6c918fc9e5c938f5aaae83c  guix-build-4f56c9145a60/output/arm-linux-gnueabihf/bitcoin-4f56c9145a60-arm-linux-gnueabihf-debug.tar.gz
0870aa09463c757f12fd27e29b55fa725a7fc62f5bc3dd02a41a99968a8faa7e  guix-build-4f56c9145a60/output/arm-linux-gnueabihf/bitcoin-4f56c9145a60-arm-linux-gnueabihf.tar.gz
222a1bee4640df2263e92fd6467f622fbdd6e15fe9d377f10916f2832ae9c0f5  guix-build-4f56c9145a60/output/arm64-apple-darwin/SHA256SUMS.part
1f669eb0faddbfeb0a2c23514e0e4653cac965e086e92076b1d8e48702268417  guix-build-4f56c9145a60/output/arm64-apple-darwin/bitcoin-4f56c9145a60-arm64-apple-darwin-codesigning.tar.gz
58978c20a84de0676ed9b1e087fb6c3a9e28ab3ad75ac269dd13dac83cebe4f5  guix-build-4f56c9145a60/output/arm64-apple-darwin/bitcoin-4f56c9145a60-arm64-apple-darwin-unsigned.tar.gz
345c1b41712de6eea8df60be38dac7c9207ac887e824f8a0802c1acc59d31493  guix-build-4f56c9145a60/output/arm64-apple-darwin/bitcoin-4f56c9145a60-arm64-apple-darwin-unsigned.zip
c3dbbe57d48d93360c78b3cb67593e3c5e8ddca7db407f1c065689db5ddcc3ec  guix-build-4f56c9145a60/output/dist-archive/bitcoin-4f56c9145a60.tar.gz
858eab2b79b7e68d6b94a13cf1273aab3932357b2430f9aca53c096d31520471  guix-build-4f56c9145a60/output/powerpc64-linux-gnu/SHA256SUMS.part
c3269ae4d281d368de79de007893addeee8f3415e5d2e745da8522eab28c1df4  guix-build-4f56c9145a60/output/powerpc64-linux-gnu/bitcoin-4f56c9145a60-powerpc64-linux-gnu-debug.tar.gz
52f6866dfd6ad777f8301d9d81e24571c35df3be6836fe5d05eaf07f10902cef  guix-build-4f56c9145a60/output/powerpc64-linux-gnu/bitcoin-4f56c9145a60-powerpc64-linux-gnu.tar.gz
5d20762b4e3cccfda826926e14e34b7e483afd2f9229550de7a83053d3c5e31e  guix-build-4f56c9145a60/output/riscv64-linux-gnu/SHA256SUMS.part
bfc9f7b185924de453a8cc1c3af95fb2319c1815827163c1f1062fad9273f12c  guix-build-4f56c9145a60/output/riscv64-linux-gnu/bitcoin-4f56c9145a60-riscv64-linux-gnu-debug.tar.gz
c02f74be2ac7a3ad2eca238dda9f738768bae1a8fdcd35da14ca51b86398cfa1  guix-build-4f56c9145a60/output/riscv64-linux-gnu/bitcoin-4f56c9145a60-riscv64-linux-gnu.tar.gz
42b46aec307e240b7e0f560b499e5e8a10f34287618a468366b725d74c291199  guix-build-4f56c9145a60/output/x86_64-apple-darwin/SHA256SUMS.part
aa616dffc9d2e7ba049f80c81fe0b39f01cd9d9fc159634841aefeb42480aa7e  guix-build-4f56c9145a60/output/x86_64-apple-darwin/bitcoin-4f56c9145a60-x86_64-apple-darwin-codesigning.tar.gz
3dca89a7ff073d436e34a6a1230ae280db611030d2441d33b733c41c9da72c1b  guix-build-4f56c9145a60/output/x86_64-apple-darwin/bitcoin-4f56c9145a60-x86_64-apple-darwin-unsigned.tar.gz
83a58e7613671f7a6e6f92115cfd356520623222bd408b0f8360d25b1972d3f6  guix-build-4f56c9145a60/output/x86_64-apple-darwin/bitcoin-4f56c9145a60-x86_64-apple-darwin-unsigned.zip
974ee6053b728441f1347202fd824aea1a2179671fad98b810a817a57024f3e2  guix-build-4f56c9145a60/output/x86_64-linux-gnu/SHA256SUMS.part
da8ad098903d32a9f9e92a0b54d9fc9f17453f4cfb41c67fd8fac7ef65f818d1  guix-build-4f56c9145a60/output/x86_64-linux-gnu/bitcoin-4f56c9145a60-x86_64-linux-gnu-debug.tar.gz
ba538c4782a5a0e03b7c01d462117dfe7346b52b674a0f8a356e0eba65fc2fa3  guix-build-4f56c9145a60/output/x86_64-linux-gnu/bitcoin-4f56c9145a60-x86_64-linux-gnu.tar.gz
1c48ddb35faf7cd75d5f1848b1ef9dd074e4216dbe6238d9ee84fa8a5050ca63  guix-build-4f56c9145a60/output/x86_64-w64-mingw32/SHA256SUMS.part
065ec2d0ae35c218a8f1c2c821aeb3054ac825c83f0de149516b147a7590a50e  guix-build-4f56c9145a60/output/x86_64-w64-mingw32/bitcoin-4f56c9145a60-win64-codesigning.tar.gz
f2550d7d7f17b1267a9eacbbd7f1fafd098a76abf39236f041f1985bc38086c4  guix-build-4f56c9145a60/output/x86_64-w64-mingw32/bitcoin-4f56c9145a60-win64-debug.zip
6e6dfca1c2079ab7da9e7b036c11f34cacec18292b7d96f269557dddc7e66b72  guix-build-4f56c9145a60/output/x86_64-w64-mingw32/bitcoin-4f56c9145a60-win64-setup-unsigned.exe
b5895e8652486f54009a6638a64b91bd546d65a8282d841ee1ee28986894d983  guix-build-4f56c9145a60/output/x86_64-w64-mingw32/bitcoin-4f56c9145a60-win64-unsigned.zip
```

fanquake reviewed (approved):
ACK 4f56c9145a60c4fb837f11e47c5aa39ad8fa3523

hebasto commented:
My Guix build:
```
aarch64
8ee8f51264bd5f28d838207486f83db69a1dd4d00eeda009f35ef9ebff794d0f  guix-build-4f56c9145a60/output/aarch64-linux-gnu/SHA256SUMS.part
aaf5a45736f062df340bbd434295a087f90897401a8d9e1489f5ad56c4a82fab  guix-build-4f56c9145a60/output/aarch64-linux-gnu/bitcoin-4f56c9145a60-aarch64-linux-gnu-debug.tar.gz
d6b4a08a7e9dad84dcfdc504896cf0b13a4c2c416ef8608903cd542297936f74  guix-build-4f56c9145a60/output/aarch64-linux-gnu/bitcoin-4f56c9145a60-aarch64-linux-gnu.tar.gz
c236ea00cde9621315a84153c0e672c8f9299d93efe379ff82188e2e018167e3  guix-build-4f56c9145a60/output/arm-linux-gnueabihf/SHA256SUMS.part
719c221a599298d5795ce82c0dc051ea512e2f17d6c918fc9e5c938f5aaae83c  guix-build-4f56c9145a60/output/arm-linux-gnueabihf/bitcoin-4f56c9145a60-arm-linux-gnueabihf-debug.tar.gz
0870aa09463c757f12fd27e29b55fa725a7fc62f5bc3dd02a41a99968a8faa7e  guix-build-4f56c9145a60/output/arm-linux-gnueabihf/bitcoin-4f56c9145a60-arm-linux-gnueabihf.tar.gz
222a1bee4640df2263e92fd6467f622fbdd6e15fe9d377f10916f2832ae9c0f5  guix-build-4f56c9145a60/output/arm64-apple-darwin/SHA256SUMS.part
1f669eb0faddbfeb0a2c23514e0e4653cac965e086e92076b1d8e48702268417  guix-build-4f56c9145a60/output/arm64-apple-darwin/bitcoin-4f56c9145a60-arm64-apple-darwin-codesigning.tar.gz
58978c20a84de0676ed9b1e087fb6c3a9e28ab3ad75ac269dd13dac83cebe4f5  guix-build-4f56c9145a60/output/arm64-apple-darwin/bitcoin-4f56c9145a60-arm64-apple-darwin-unsigned.tar.gz
345c1b41712de6eea8df60be38dac7c9207ac887e824f8a0802c1acc59d31493  guix-build-4f56c9145a60/output/arm64-apple-darwin/bitcoin-4f56c9145a60-arm64-apple-darwin-unsigned.zip
c3dbbe57d48d93360c78b3cb67593e3c5e8ddca7db407f1c065689db5ddcc3ec  guix-build-4f56c9145a60/output/dist-archive/bitcoin-4f56c9145a60.tar.gz
858eab2b79b7e68d6b94a13cf1273aab3932357b2430f9aca53c096d31520471  guix-build-4f56c9145a60/output/powerpc64-linux-gnu/SHA256SUMS.part
c3269ae4d281d368de79de007893addeee8f3415e5d2e745da8522eab28c1df4  guix-build-4f56c9145a60/output/powerpc64-linux-gnu/bitcoin-4f56c9145a60-powerpc64-linux-gnu-debug.tar.gz
52f6866dfd6ad777f8301d9d81e24571c35df3be6836fe5d05eaf07f10902cef  guix-build-4f56c9145a60/output/powerpc64-linux-gnu/bitcoin-4f56c9145a60-powerpc64-linux-gnu.tar.gz
5d20762b4e3cccfda826926e14e34b7e483afd2f9229550de7a83053d3c5e31e  guix-build-4f56c9145a60/output/riscv64-linux-gnu/SHA256SUMS.part
bfc9f7b185924de453a8cc1c3af95fb2319c1815827163c1f1062fad9273f12c  guix-build-4f56c9145a60/output/riscv64-linux-gnu/bitcoin-4f56c9145a60-riscv64-linux-gnu-debug.tar.gz
c02f74be2ac7a3ad2eca238dda9f738768bae1a8fdcd35da14ca51b86398cfa1  guix-build-4f56c9145a60/output/riscv64-linux-gnu/bitcoin-4f56c9145a60-riscv64-linux-gnu.tar.gz
42b46aec307e240b7e0f560b499e5e8a10f34287618a468366b725d74c291199  guix-build-4f56c9145a60/output/x86_64-apple-darwin/SHA256SUMS.part
aa616dffc9d2e7ba049f80c81fe0b39f01cd9d9fc159634841aefeb42480aa7e  guix-build-4f56c9145a60/output/x86_64-apple-darwin/bitcoin-4f56c9145a60-x86_64-apple-darwin-codesigning.tar.gz
3dca89a7ff073d436e34a6a1230ae280db611030d2441d33b733c41c9da72c1b  guix-build-4f56c9145a60/output/x86_64-apple-darwin/bitcoin-4f56c9145a60-x86_64-apple-darwin-unsigned.tar.gz
83a58e7613671f7a6e6f92115cfd356520623222bd408b0f8360d25b1972d3f6  guix-build-4f56c9145a60/output/x86_64-apple-darwin/bitcoin-4f56c9145a60-x86_64-apple-darwin-unsigned.zip
974ee6053b728441f1347202fd824aea1a2179671fad98b810a817a57024f3e2  guix-build-4f56c9145a60/output/x86_64-linux-gnu/SHA256SUMS.part
da8ad098903d32a9f9e92a0b54d9fc9f17453f4cfb41c67fd8fac7ef65f818d1  guix-build-4f56c9145a60/output/x86_64-linux-gnu/bitcoin-4f56c9145a60-x86_64-linux-gnu-debug.tar.gz
ba538c4782a5a0e03b7c01d462117dfe7346b52b674a0f8a356e0eba65fc2fa3  guix-build-4f56c9145a60/output/x86_64-linux-gnu/bitcoin-4f56c9145a60-x86_64-linux-gnu.tar.gz
1c48ddb35faf7cd75d5f1848b1ef9dd074e4216dbe6238d9ee84fa8a5050ca63  guix-build-4f56c9145a60/output/x86_64-w64-mingw32/SHA256SUMS.part
065ec2d0ae35c218a8f1c2c821aeb3054ac825c83f0de149516b147a7590a50e  guix-build-4f56c9145a60/output/x86_64-w64-mingw32/bitcoin-4f56c9145a60-win64-codesigning.tar.gz
f2550d7d7f17b1267a9eacbbd7f1fafd098a76abf39236f041f1985bc38086c4  guix-build-4f56c9145a60/output/x86_64-w64-mingw32/bitcoin-4f56c9145a60-win64-debug.zip
6e6dfca1c2079ab7da9e7b036c11f34cacec18292b7d96f269557dddc7e66b72  guix-build-4f56c9145a60/output/x86_64-w64-mingw32/bitcoin-4f56c9145a60-win64-setup-unsigned.exe
b5895e8652486f54009a6638a64b91bd546d65a8282d841ee1ee28986894d983  guix-build-4f56c9145a60/output/x86_64-w64-mingw32/bitcoin-4f56c9145a60-win64-unsigned.zip
```

hebasto reviewed (commented):
Post-merge re-ACK 4f56c9145a60c4fb837f11e47c5aa39ad8fa3523.

fanquake commented:
Looks like this is getting pulled in by `python-scikit-build-core` -> `"/gnu/store/8a28mn0n9bd0nwq5vd6d1lfhr1j27cbk-rust-ring-0.17.8.tar.gz.drv"` -> `"/gnu/store/wbfmclxgnmlqbin55ipb0x05i1yapiy5-go-1.21.13.drv"` -> `"/gnu/store/wb36d09msn72aamvj8hj4y55551rqilz-go-1.17.13.drv"` -> `"/gnu/store/mjvb787fwd5bb6m2z02hhy30n5jzd9gw-go-1.4-bootstrap-20171003.drv"`

Sjors commented:
This is also causing (different) issues on my second guix machine: #32759

Sjors commented on contrib/guix/manifest.scm:
```diff
@@ -15,8 +15,10 @@
+             ((gnu packages python-build) #:select (python-poetry-core))
              ((gnu packages python-crypto) #:select (python-asn1crypto))
+             ((gnu packages python-science) #:select (python-scikit-build-core))
+             ((gnu packages python-xyz) #:select (python-pydantic-2 python-pydantic-core))
```
765922d8022d3addeb80b5f3f6e041f7fd2ad3ed: do you know why https://codeberg.org/guix/guix/pulls/353/files only needs `python-pydantic-2` and not `python-pydantic-core`?
